observability/src/main/java/io/grpc/observability/interceptors/GcpLogHelper.java

```diff
@@ -0,0 +1,430 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.observability.interceptors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Charsets;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Duration;
+import com.google.protobuf.util.Timestamps;
+import io.grpc.Attributes;
+import io.grpc.Deadline;
+import io.grpc.Grpc;
+import io.grpc.InternalMetadata;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor.Marshaller;
+import io.grpc.Status;
+import io.grpc.internal.TimeProvider;
+import io.grpc.observability.logging.Sink;
+import io.grpc.observabilitylog.v1.GrpcLogRecord;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.Address;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.EventType;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.LogLevel;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Helper class for GCP observability logging.
+ */
+final class GcpLogHelper {
+  private static final Logger logger = Logger.getLogger(GcpLogHelper.class.getName());
+  static final Metadata.Key<byte[]> STATUS_DETAILS_KEY =
+      Metadata.Key.of(
+          "grpc-status-details-bin",
+          Metadata.BINARY_BYTE_MARSHALLER);
+
+  /**
+   *  Class for proto related helper methods.
+   */
+  static final class GcpLogSinkWriter {
+    private final Sink sink;
+    private final TimeProvider timeProvider;
+
+    GcpLogSinkWriter(Sink sink, TimeProvider timeProvider) {
+      this.sink = sink;
+      this.timeProvider = timeProvider;
+    }
+
+    GrpcLogRecord.Builder createTimestamp() {
+      long nanos = timeProvider.currentTimeNanos();
+      return GrpcLogRecord.newBuilder().setTimestamp(Timestamps.fromNanos(nanos));
+    }
+
+
+    /**
+     * Logs the client header.
+     */
+    void logClientHeader(
+        long seqId,
+        String serviceName,
+        String methodName,
+        @Nullable String authority,
+        @Nullable Duration timeout,
+        Metadata metadata,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId,
+        @Nullable SocketAddress peerAddress) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(rpcId, "rpcId");
+      checkArgument(
```

Comment:
Something to think about: if any of these checks fails it will throw an exception (NPR or IllegalArgumentException) and abort the RPC. Of course serviceName etc cannot be null so it looks okay but the implication is that enabling observability/logging has resulted into some RPCs getting aborted. So pls doubly make sure that these checks check some basic invariants